conans/client/remote_manager.py

```diff
@@ -24,20 +25,63 @@ def upload_conan(self, conan_reference, remote):
         """Will upload the conans to the first remote"""
         basedir = self._client_cache.export(conan_reference)
         rel_files = self._client_cache.export_paths(conan_reference)
+        the_files = {filename: os.path.join(basedir, filename) for filename in rel_files}
 
-        the_files = build_files_set(basedir, rel_files)
-        the_files = compress_export_files(the_files)
+        if CONANFILE not in rel_files or CONAN_MANIFEST not in rel_files:
+            raise ConanException("Cannot upload corrupted recipe '%s'" % str(conan_reference))
+
+        # FIXME: Check modified exports by hand?
+
+        if EXPORT_TGZ_NAME not in the_files:
+            self._output.rewrite_line("Compressing exported files...")
+            the_files = compress_export_files(the_files, basedir)
+        else:
+            the_files = {EXPORT_TGZ_NAME: the_files[EXPORT_TGZ_NAME],
+                         CONANFILE: the_files[CONANFILE],
+                         CONAN_MANIFEST: the_files[CONAN_MANIFEST]}
         return self._call_remote(remote, "upload_conan", conan_reference, the_files)
 
     def upload_package(self, package_reference, remote):
         """Will upload the package to the first remote"""
+        t1 = time.time()
         basedir = self._client_cache.package(package_reference)
         rel_files = self._client_cache.package_paths(package_reference)
 
-        the_files = build_files_set(basedir, rel_files)
-        self._output.rewrite_line("Compressing package...")
-        the_files = compress_package_files(the_files)
-        return self._call_remote(remote, "upload_package", package_reference, the_files)
+        self._output.rewrite_line("Checking package integrity...")
+        if CONANINFO not in rel_files or CONAN_MANIFEST not in rel_files:
+            raise ConanException("Cannot upload corrupted package '%s'" % str(package_reference))
+
+        the_files = {filename: os.path.join(basedir, filename) for filename in rel_files}
+        logger.debug("====> Time remote_manager build_files_set : %f" % (time.time() - t1))
+
+        # If package has been modified remove tgz to regenerate it
+        read_manifest, expected_manifest = self._client_cache.package_manifests(package_reference)
+        if read_manifest is None or read_manifest.file_sums != expected_manifest.file_sums:
+            if PACKAGE_TGZ_NAME in the_files:
+                del the_files[PACKAGE_TGZ_NAME]
```

Comment:
Why bothering delete this from dict, if ``ConanException`` will be raised and processing stopped?